### PR TITLE
Summary row

### PR DESCRIPTION
**Pull request by SirWojtek**
**What kind of change does this PR introduce?** (check one with "x")
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/swimlane/ngx-datatable/issues/1215

**What is the new behavior?**
https://github.com/swimlane/ngx-datatable/issues/1215

**Does this PR introduce a breaking change?** (check one with "x")
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
The full specification of the feature can be found on the linked issue.

## TODO list:
- [x] skeleton of the solution
- [x] simple demo page
- [x] material styles
- [x] boostrap styles
- [x] dark styles
- [x] basic summary row
- [x] custom summary function
- [x] summary position
- [x] custom summary cell template
- [x] regression check (pagination, sort, styles)
- [x] more demos
- [x] width bug
- [x] tests

### CODE REVIEW

**Comment by amcdnl** [comment]
Looks good overall :)

**Comment by maciej-karpus** [comment]
Hi @SirWojtek, thank you for undertaking this task. Could you give us an approximate time of completion?

**Comment by SirWojtek** [comment]
Hello, I'll try to finish it on Friday (19.01.2018) but I cannot guarantee that. In the worst case, PR will be ready before the end of next week.

**Comment by SirWojtek** [comment]
I found interesting style bug in case of a table with summary row + server-side paging.
### Right after load page
width-bug
### After changing window width
width-bug2

I guess it has something with delayed data loading and recalculating/refreshing layout. I've checked the width of summary cells and they had the proper inline width set in both cases. Any clues?

**Comment by amcdnl** [comment]
Change detection not firing correctly would be my assumption.

**Comment by SirWojtek** [comment]
OK, the change is ready to be merged if you don't have any comments.

**Comment by startswithaj** [comment]
+1

**Comment by maciej-karpus** [comment]
+1

**Comment by maciej-karpus** [comment]
@amcdnl Hi, could you tell us when do you plan to merge this request?

**Comment by Chris-V** [comment]
How is it going with this PR? Need any help? This is something that interests us very much. Thank you!

**Comment by SirWojtek** [comment]
Hello @Chris-V, the change is ready to be merged.
I'm going to add support for HTML inline `summaryTemplate` (for more details see https://github.com/swimlane/ngx-datatable/issues/1215#issuecomment-364889868 and following) but I can make it in a next PR.

**Comment by Chris-V** [comment]
That would be appreciated. We will need `summaryTemplate`s eventually. In the mean time the function will be great.

**Comment by Chris-V** [comment]
@SirWojtek Rather than wait for it to be merged, I went ahead and made my own fork. Hopefully it gets merged sooner than later.

In the meantime, I have two small concerns given our usage
1. Anyways, I noticed that the `summaryFunc` is applied even though I provide a `summaryTemplate`. This causes issues when the data cannot be reduced properly with the default function (ie: I'm getting an error "Reduce of empty array with no initial value" on more complex cells).
    As a workaround, I provide a noop function to `summaryFunc`.
2. We must display complex data in the summary row, often spanning more than one line. It would be great being able to set a different rowHeight for the summary row alone.

Btw, thanks for the great work, this is something very useful for our project.

**Comment by SirWojtek** [comment]
Thank you for the feedback! I'll investigate that in next week. Cheers 👌

**Comment by SirWojtek** [comment]
Hi @Chris-V, I've prepared fixes for issues you mentioned.

**Comment by maciej-karpus** [comment]
Hi guys, could you give me advice how to use your changes in my package.json?

**Comment by Chris-V** [comment]
@SirWojtek Thanks a lot! This is working perfectly :D

@maciej-karpus You need to add a git repository with the compiled code in your package.json
1. Fork the project
1. Apply this PR to your fork. You can append `.patch` to the PR's URL to get the patch file, ie: https://github.com/swimlane/ngx-datatable/pull/1233.patch
1. Build the project `npm run package` and commit that to your fork. You can also tag this commit to make things easier to track.

You can use my fork to try the feature, but be warned that I won't maintain it. Add this to your package.json's dependencies: 
```
"@swimlane/ngx-datatable": "git://github.com/ArcBees/ngx-datatable.git#arcbees.1"
```

**Comment by maciej-karpus** [comment]
@Chris-V Thank you for the instruction and your fork. :)

**Comment by patrickreck** [comment]
What are we waiting for on this one?

**Comment by SirWojtek** [comment]
Unfortunately, I cannot merge that PR by myself. I'm waiting for someone with merge privileges to do that but seems they are inactive. Maybe it's time to fork off and mark this repository as unmaintained?

**Comment by marjan-georgiev** [comment]
Looks good! Thanks @SirWojtek! Will release shortly 👍

File: src/components/body/summary/summary-row.component.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    this.columns
+      .filter(col => !col.summaryTemplate)
+      .forEach(col => {
+      const cellsFromSingleColumn = this.rows.map(row => row[col.prop]);
+      const sumFunc = col.summaryFunc || defaultSumFunc;
+
+      this.summaryRow[col.prop] = col.pipe ?
+        col.pipe.transform(sumFunc(cellsFromSingleColumn)) :
+        sumFunc(cellsFromSingleColumn);
```
Comment (tHe-AK) [inline src/components/body/summary/summary-row.component.ts:71]: @SirWojtek Can we pass 'col' in this method as I have a single summaryFunc with minor changes for different columns?

Comment (SirWojtek) [reply src/components/body/summary/summary-row.component.ts:71]: Hey @tHe-AK, can I ask you to provide your use case? I'd like to see your summaryFunc and how you inject it to datatable.

Comment (tHe-AK) [reply src/components/body/summary/summary-row.component.ts:71]: Use Case:
Want to show a fixed row between header & data-rows which should remain there for all pages & filter results. This row is not showing any total of the cells. I have to show some fixed values stored in an object.

Summary function i want to use:
summaryFunc = (cells**, col**) => {
    return myObj[col.prop];
}

**Pull request merged**